### PR TITLE
feat: Introduce FileNameFactory class for the settlement method job

### PR DESCRIPTION
**Pull request by JesperGomes**
# Description

## Pull-request quality

- [x] The title adheres to [this guide](https://github.com/Mech0z/GitHubGuidelines)
- [x] Tests are written and executed locally
- [ ] Subsystem tests have been tested (by manually deploying to `dev_002`)

### CODE REVIEW

File: source/databricks/settlement_report/settlement_report_job/domain/report_data_type.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from enum import Enum
+
+
+class ReportDataType(Enum):
```
Comment (BjarkeMeier) [inline source/databricks/settlement_report/settlement_report_job/domain/report_data_type.py:18]: What's the purpose? What happens if I change the values? Can we add a little docstring?

File: source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
+from collections.abc import Iterable
+from zoneinfo import ZoneInfo
+
+from settlement_report_job.domain.market_role import MarketRole
+from settlement_report_job.domain.report_data_type import ReportDataType
+from settlement_report_job.domain.settlement_report_args import SettlementReportArgs
+
+
+class MarketRoleInFileName:
```
Comment (BjarkeMeier) [inline source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py:9]: The market role in the file name or the name of the market role in the file?

Comment (JesperGomes) [reply source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py:9]: I will add a doc string :-)

File: source/databricks/settlement_report/tests/domain/test_report_name_factory.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+from settlement_report_job.domain.market_role import MarketRole
+from settlement_report_job.domain.report_name_factory import (
+    FileNameFactory,
+    ReportDataType,
+)
+from settlement_report_job.domain.settlement_report_args import SettlementReportArgs
+
+
+@pytest.fixture(scope="session")
+def default_settlement_report_args() -> SettlementReportArgs:
```
Comment (BjarkeMeier) [inline source/databricks/settlement_report/tests/domain/test_report_name_factory.py:17]: How about moving this fixture to tests/conftest.py?

File: source/databricks/settlement_report/tests/domain/test_report_name_factory.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    energy_supplier = "1234567890123"
+    grid_area_code = "123"
+    sut = FileNameFactory(report_data_type, default_settlement_report_args)
+
+    # Act
+    actual = sut.create(grid_area_code, energy_supplier, chunk_index=None)
+
+    # Assert
+    assert (
+        actual == f"{expected_pre_fix}_123_1234567890123_DDQ_01-07-2024_01-08-2024.csv"
```
Comment (BjarkeMeier) [inline source/databricks/settlement_report/tests/domain/test_report_name_factory.py:62]:
```suggestion
        actual == f"{expected_pre_fix}_{grid_area_code}_{energy_supplier_id}_DDQ_01-07-2024_01-08-2024.csv"
```

Comment (BjarkeMeier) [inline source/databricks/settlement_report/tests/domain/test_report_name_factory.py:62]: 01_08? Why not 31_07?

Comment (JesperGomes) [reply source/databricks/settlement_report/tests/domain/test_report_name_factory.py:62]: Oh - good question. I think you are right

File: source/databricks/settlement_report/tests/domain/test_report_name_factory.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    grid_area_code = "123"
+    sut = FileNameFactory(
+        ReportDataType.TimeSeriesHourly, default_settlement_report_args
+    )
+
+    # Act
+    actual = sut.create(grid_area_code, energy_supplier_id=None, chunk_index=None)
+
+    # Assert
+    assert actual == "TSSD60_123_1111111111111_DDM_01-07-2024_01-08-2024.csv"
```
Comment (BjarkeMeier) [inline source/databricks/settlement_report/tests/domain/test_report_name_factory.py:84]: Use substitution variables?

File: source/databricks/settlement_report/tests/domain/test_report_name_factory.py (hunk 3 above)
Comment (BjarkeMeier) [inline source/databricks/settlement_report/tests/domain/test_report_name_factory.py:17]: Perhaps add a docstring with the period values as you assume certain period values in your tests.

Comment (JesperGomes) [reply source/databricks/settlement_report/tests/domain/test_report_name_factory.py:17]: I'm not sure it is worth making the coupling between the tests!?!

File: source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    """
+    Market role identifiers used in the csv file name in the settlement report.
+    This is not the role of the requesting actor, but the role of the actor that the data is related to.
```
Comment (BjarkeMeier) [inline source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py:12]: Can you request on behalf of someone else?

File: source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            ReportDataType.TimeSeriesHourly,
+            ReportDataType.TimeSeriesQuarterly,
+        }:
+            return self._create_time_series_filename(
+                grid_area_code, energy_supplier_id, chunk_index
+            )
+        else:
+            raise NotImplementedError(
+                f"Report data type {self.report_data_type} is not supported."
+            )
```
Comment (BjarkeMeier) [inline source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py:40]: This pattern is better (exit early and reduce nesting). Now the code is written in the style of guards/pre-conditions.

```suggestion
        if self.report_data_type not in {
            ReportDataType.TimeSeriesHourly,
            ReportDataType.TimeSeriesQuarterly,
        }:
            raise NotImplementedError(
                f"Report data type {self.report_data_type} is not supported."
            )

        return self._create_time_series_filename(
            grid_area_code, energy_supplier_id, chunk_index
        )
```

File: source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        time_zone_info = ZoneInfo(self.args.time_zone)
+        return (
+            self.args.period_end.astimezone(time_zone_info) - timedelta(days=1)
+        ).strftime("%d-%m-%Y")
+
+    def _get_pre_fix(self) -> str:
+        if self.report_data_type == ReportDataType.TimeSeriesHourly:
+            return "TSSD60"
+        elif self.report_data_type == ReportDataType.TimeSeriesQuarterly:
+            return "TSSD15"
```
Comment (BjarkeMeier) [inline source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py:88]: Should we add a final "else"?

File: source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py (hunk 6 above)
Comment (JesperGomes) [reply source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py:12]: yes

File: source/databricks/settlement_report/tests/domain/test_report_name_factory.py (hunk 3 above)
Comment (BjarkeMeier) [reply source/databricks/settlement_report/tests/domain/test_report_name_factory.py:17]: docstrings and comments are distinct concepts. A docstring will - in many editors - provide pop-up documentation for the developer when hovering usages.

**Review by BjarkeMeier — APPROVED** [review]

File: source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py (hunk 7 above)
Comment (JesperGomes) [reply source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py:40]: then we will soon have an extra statement. if not in {   } in the beginning and then all the if, elif, elif. I'm not sure that is better?

Comment (BjarkeMeier) [reply source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py:40]: Oh, sorry. Then let's keep it.

File: source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py (hunk 8 above)
Comment (JesperGomes) [reply source/databricks/settlement_report/settlement_report_job/domain/report_name_factory.py:88]: there is a guard in the create method. We can also raise an exception here if you prefer

**Review by BjarkeMeier — DISMISSED** [review]

**Review by BjarkeMeier — APPROVED** [review]

**Pull request merged**